src/provenance_gc.rs

```diff
@@ -0,0 +1,191 @@
+    }
+}
+
+impl VisitProvenance for Pointer<Option<Provenance>> {
+    fn visit_provenance(&self, visit: &mut VisitWith<'_>) {
+        let (prov, _offset) = self.into_parts();
+        prov.visit_provenance(visit);
+    }
+}
+
+impl VisitProvenance for Scalar<Provenance> {
+    fn visit_provenance(&self, visit: &mut VisitWith<'_>) {
+        match self {
+            Scalar::Ptr(ptr, _) => ptr.visit_provenance(visit),
+            Scalar::Int(_) => (),
+        }
+    }
+}
+
+impl VisitProvenance for Immediate<Provenance> {
+    fn visit_provenance(&self, visit: &mut VisitWith<'_>) {
+        match self {
+            Immediate::Scalar(s) => {
+                s.visit_provenance(visit);
+            }
+            Immediate::ScalarPair(s1, s2) => {
+                s1.visit_provenance(visit);
+                s2.visit_provenance(visit);
+            }
+            Immediate::Uninit => {}
+        }
+    }
+}
+
+impl VisitProvenance for MemPlaceMeta<Provenance> {
+    fn visit_provenance(&self, visit: &mut VisitWith<'_>) {
+        match self {
+            MemPlaceMeta::Meta(m) => m.visit_provenance(visit),
+            MemPlaceMeta::None => {}
+        }
+    }
+}
+
+impl VisitProvenance for ImmTy<'_, Provenance> {
+    fn visit_provenance(&self, visit: &mut VisitWith<'_>) {
+        (**self).visit_provenance(visit)
+    }
+}
+
+impl VisitProvenance for MPlaceTy<'_, Provenance> {
+    fn visit_provenance(&self, visit: &mut VisitWith<'_>) {
+        self.ptr().visit_provenance(visit);
+        self.meta().visit_provenance(visit);
+    }
+}
+
+impl VisitProvenance for PlaceTy<'_, Provenance> {
+    fn visit_provenance(&self, visit: &mut VisitWith<'_>) {
+        match self.as_mplace_or_local() {
+            Either::Left(mplace) => mplace.visit_provenance(visit),
+            Either::Right(_) => (),
+        }
+    }
+}
+
+impl VisitProvenance for OpTy<'_, Provenance> {
+    fn visit_provenance(&self, visit: &mut VisitWith<'_>) {
+        match self.as_mplace_or_imm() {
+            Either::Left(mplace) => mplace.visit_provenance(visit),
+            Either::Right(imm) => imm.visit_provenance(visit),
+        }
+    }
+}
+
+impl VisitProvenance for Allocation<Provenance, AllocExtra<'_>> {
+    fn visit_provenance(&self, visit: &mut VisitWith<'_>) {
+        for prov in self.provenance().provenances() {
+            prov.visit_provenance(visit);
+        }
+
+        self.extra.visit_provenance(visit);
+    }
+}
+
+impl VisitProvenance for crate::MiriInterpCx<'_, '_> {
+    fn visit_provenance(&self, visit: &mut VisitWith<'_>) {
+        // Visit the contents of the allocations and the IDs themselves, to account for all
+        // live allocation IDs and all provenance in the allocation bytes, even if they are leaked.
+        // Here we exploit that `adjust_allocation` always returns `Owned`, to all
+        // tcx-managed allocations ever read or written will be copied in `alloc_map`.
+        self.memory.alloc_map().iter(|it| {
```

Comment:
We also have the secondary memory map in `tcx`. I guess we have to walk that one as well, judging from the test failures. The comment I suggested about not having to worry about tcx-managed allocation seems to be wrong.

In particular it is wrong for function and vtable allocations, which never get copied into the local `alloc_map`.